mkdocs/__main__.py

```diff
@@ -4,28 +4,62 @@
 import sys
 import logging
 import click
+import textwrap
 
 from mkdocs import __version__
 from mkdocs import utils
-from mkdocs import exceptions
 from mkdocs import config
 from mkdocs.commands import build, gh_deploy, new, serve
 
 log = logging.getLogger(__name__)
 
 
+class ColorFormatter(logging.Formatter):
+    colors = {
+        'CRITICAL': 'red',
+        'ERROR': 'red',
+        'WARNING': 'yellow',
+        'DEBUG': 'blue'
+    }
+
+    def format(self, record):
+        prefix = f'{record.levelname:<8} -  '
+        if record.levelname in self.colors:
+            prefix = click.style(prefix, fg=self.colors[record.levelname])
+        lines = textwrap.wrap(record.getMessage(), width=68)
```

Comment:
Why wrap, *and* with this super specific choice of width? This looks horrible.

![image](https://user-images.githubusercontent.com/371383/115142156-81257980-a040-11eb-8f45-ec34a033e14a.png)

Reply:
> This looks horrible.

That is a matter of opinion. Regardless, I chose that width because I still run my terminal with a width of 80. With the indentation of 12, that leaves a width of 68. It is much easier to read with the indentation than without.

Reply:
In this example not a matter of opinion because the wrapping is literally incorrect and it's very hard to deduce what this was supposed to be.

But if generally it is a matter of opinion, then indeed, why should you enforce your opinion?

Can you name me another tool that wraps its output at 68 characters? Especially in logs???

For reference, my terminal is usually around 230 characters wide.

Reply:
You seem to be missing the purpose of this. So let me illustrate with the screenshot below.

![Screen Shot 2021-04-18 at 5 26 22 PM](https://user-images.githubusercontent.com/78846/115161409-7d5b1c80-a06b-11eb-9d3c-e2799755c786.png)

The first `build` was run with the text wrap enabled. The second was with it disabled. Notice how much easier it is to read the first. The difference is that the left side is indented on secondary lines. I suppose if you have a wider terminal window that may not be so evident. But it was a huge improvement for me. 

If you have a better way to get the same effect, I'm open to suggestions. But barring anything better, the current behavior is an improvement over what we had.

I don't know, but is there an easy way to obtain the terminal width from Python? If so we could wrap based on that and no-one would get what they perceive to be weird results.

Reply:
> You seem to be missing the purpose of this.

No of course I'm not missing it. I just don't want it, it's a downgrade in almost every case.

It changes from an unambiguous representation to an ambiguous one. (see my example)

It breaks line-based searches (yes, I have done `grep` on the errors) and line counts and just everything.

Other tools just don't do this!

And why does the 80 char terminal suddenly become so important *now* after so many years. Did anyone else at all mention this?

Reply:
> And why does the 80 char terminal suddenly become so important now after so many years. Did anyone else at all mention this?

It's not 80 chars specifically, but the general motivation for all of the changes here are related to #1577 and #1888. As you can see, those two requests are in opposition to each other. In the case of #1888, the primary issue is that the user is misunderstanding the logs. So my goal was to try to make the logs more clear and easier to read. The color certainly helps by better differentiating between warnings and informational messages. But, at the same time, new users often simply see a wall of text and don't try to understand it. By providing some formatting (indenting secondary lines), it makes it easier to read However, the only way to indent the lines is to also hard-wrap them.

Truth be told,  when I set out I didn't have text wrapping in mind at all. However, as I was working on it, it occurred to me that it might help and I tried it. I was surprised how much easier it made the logs to read, so I moved on and never gave it another thought.

Things like greping logs never occurred to me. After all, the formatter is specifically on the StreamHandler which writes to the terminal. If someone is writing to a log file, there technically should be a different handler which does that. True, a user could be redirecting std.out/std.error to a file, but that means the user is most likely using the server in something other than a local dev environment, which we have never officially supported. In that case, you should be using a different server anyway.

My point is that the only use case that we officially support is to have the logs written directly to the terminal. In that case there is little harm in splitting lines (I suppose you could grep the output of `mkdocs build`, but who has that much output from a single build?). And if I had thought to call `shutil.get_terminal_size` to calculate line length, you may have never even noticed so long as you are using the server as it is officially supported.

To be clear, I am not opposed to removing the textwrapping for a good reason. I just haven't seen one that fits into the officially supported use case. In my view, the change is such a huge improvement that I'm okay with leaving it despite it being a little unconventional. 

As a compromise, perhaps we could use `shutil.get_terminal_size` with a `fallback` of `None` (or `(0, 0)` or some other obviously invalid value). That way, if `shutil.get_terminal_size` returns the fallback, then we can assume there is no "terminal" and skip the wrapping. However, if a valid value is returned, we can assume the user is on the terminal and then we can wrap the text for readability using the column width. Of course, this assumes `shutil.get_terminal_size` isn't able to obtain a terminal size when MkDocs is run as a system service or from a docker instance or whatever other unsupported environments users may be using. I may be wrong about that assumption, in which case, relying on the fallback won't help. That may require some investigation.

Reply:
> the general motivation for all of the changes here are related to

OK but those are only tangential points for your motivation, not someone actually asking for (or even mentioning) that.

> In the case of #1888, the primary issue is that the user is misunderstanding the logs

Hm? I don't see any message there that can be interpreted that way.

> the user is most likely using the server in something other than a local dev environment, which we have never officially supported

> I suppose you could grep the output of mkdocs build, but who has that much output from a single build?

Please let's not turn this into "you're holding it wrong".

> In that case, you should be using a different server anyway.

Took me a while to realize that you mean `mkdocs serve`. But that's not a strong point, as all of this applies to `mkdocs build` too.

Reply:
> However, if a valid value is returned, we can assume the user is on the terminal and then we can wrap the text for readability using the column width.

Yes, this approach would probably be (barely) acceptable to me.

```shell
$ python3 -c 'import shutil; print(shutil.get_terminal_size(fallback=(0, 0)))'      
os.terminal_size(columns=195, lines=42)

$ python3 -c 'import shutil; print(shutil.get_terminal_size(fallback=(0, 0)))' | cat
os.terminal_size(columns=0, lines=0)
```

Even so, I still think the `textwrap` implementation is not safe enough for technical text. As you can see in my screenshot, it shows `--project-` `version=` and there's no way to tell whether that was `--project- version=` or `--project-version=`.

Additionally, I can no longer select the whole log line by triple-clicking it, and even after I select things manually, it still includes extraneous spaces that were not there.

So again, it's just a downgrade for me in any case, I really just prefer the terminal to do its own wrapping.

I literally don't know any tool that does this automated word wrapping. Can you point out one?

<table><tr><td><img src="https://user-images.githubusercontent.com/371383/115283168-4f4d0980-a14b-11eb-9b39-853b7ed37575.png" width="300"></td><td><img src="https://user-images.githubusercontent.com/371383/115281737-9d610d80-a149-11eb-8566-c166fd2bb97c.png" width="300"></td></tr></table>

There are some tools that wrap only on known already-separated tokens. Perfect example:

<table><tr><td><img src="https://user-images.githubusercontent.com/371383/115281045-e06eb100-a148-11eb-9496-051d23ae1180.png" width="400"></td></tr></table>

Note how the package names are carefully wrapped without splitting them across several lines on the dashes.
And then note how textual messages are *not* wrapped, *although* (or, as I'm convinced, *because*) clearly a lot of thought was put into this UI.

Reply:
> I literally don't know any tool that does this automated word wrapping.

MkDocs' own `--help` does. But then that is just the default behavior of [click][1], the lib we use for option parsing.

```
$ mkdocs build --help
Usage: mkdocs build [OPTIONS]

  Build the MkDocs documentation

Options:
  -c, --clean / --dirty           Remove old files from the site_dir before
                                  building (the default).

  -f, --config-file FILENAME      Provide a specific MkDocs config
  -s, --strict                    Enable strict mode. This will cause MkDocs
                                  to abort the build on any warnings.

  -t, --theme [mkdocs|readthedocs]
                                  The theme to use when building your
                                  documentation.

  --use-directory-urls / --no-directory-urls
                                  Use directory URLs when building pages (the
                                  default).

  -d, --site-dir PATH             The directory to output the result of the
                                  documentation build.

  -q, --quiet                     Silence warnings
  -v, --verbose                   Enable verbose output
  -h, --help                      Show this message and exit.
```

Interestingly, that is hardwrapped at 80. And we didn't choose that. That is click's default behavior. Now, I know `--help` messages are not logs, but this is pretty standard for `--help`. It is also where I got the idea from to begin with.

[1]: https://click.palletsprojects.com/

Reply:
I think all of these flags would be very important improvements if this implementation is here to stay.

* [replace_whitespace]=False
* [drop_whitespace]=False
* [break_long_words]=False
* **[break_on_hyphens]=False** (addresses my biggest complaint)

[replace_whitespace]: https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.replace_whitespace
[drop_whitespace]: https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.drop_whitespace
[break_long_words]: https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.break_long_words
[break_on_hyphens]: https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.break_on_hyphens

I checked that GitHub Actions also doesn't report a terminal width, so that's good from the perspective of this autodetection.
[Did it in a random repo, disregard the rest](https://github.com/oprypin/mkdocs-literate-nav/runs/2384012316?check_suite_focus=true#step:4:1)